java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -0,0 +1,290 @@
+    } else if (isParametrizedType(formalType) && isParametrizedType(argType)) {
+      List<JavaType> formalTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+      List<JavaType> argTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+      return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+    } else if (isParametrizedType(formalType) && !isParametrizedType(argType)) {
+      if (formalType.erasure() == argType || argType == symbols.nullType) {
+        List<JavaType> formalTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+        List<JavaType> fakeTypes = new ArrayList<>(formalTypeSubstitutedTypes.size());
+        for (int j = 0; j < formalTypeSubstitutedTypes.size(); j++) {
+          fakeTypes.add(symbols.objectType);
+        }
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, fakeTypes);
+        return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+      } else if (argType.isSubtypeOf(formalType.erasure()) && argType.isClass()) {
+        for (JavaType superType : ((JavaType.ClassJavaType) argType).symbol.superTypes()) {
+          if (formalType.erasure() == superType.erasure()) {
+            return inferTypeSubstitution(method, currentSubstitution, isLastParam, formalType, superType);
+          }
+        }
+      }
+    } else if (formalType.isTagged(JavaType.WILDCARD)) {
+      completeSubstitution(currentSubstitution, ((JavaType.WildCardType) formalType).bound, argType);
+    }
+    return currentSubstitution;
+  }
+
+  private static TypeSubstitution mergeTypeSubstitutions(TypeSubstitution currentSubstitution, TypeSubstitution newSubstitution) {
+    TypeSubstitution result = new TypeSubstitution();
+    for (Map.Entry<JavaType.TypeVariableJavaType, JavaType> substitution : currentSubstitution.substitutionEntries()) {
+      result.add(substitution.getKey(), substitution.getValue());
+    }
+    for (Map.Entry<JavaType.TypeVariableJavaType, JavaType> substitution : newSubstitution.substitutionEntries()) {
+      if (!result.typeVariables().contains(substitution.getKey())) {
+        result.add(substitution.getKey(), substitution.getValue());
+      }
+    }
+    return result;
+  }
+
+  private void completeSubstitution(TypeSubstitution currentSubstitution, JavaType formalType, JavaType argType) {
+    if (formalType.isTagged(JavaType.TYPEVAR) && currentSubstitution.substitutedType(formalType) == null) {
+      JavaType expectedType = argType;
+      if (expectedType.isPrimitive()) {
+        expectedType = expectedType.primitiveWrapperType;
+      } else if (expectedType == symbols.nullType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)